src/capi.rs

```diff
@@ -78,6 +94,53 @@ pub struct Config {
     cfg: rav1e::Config,
 }
 
+trait EncoderContext {
+  fn new_frame(&self) -> Frame;
+  fn send_frame(&mut self, frame: Option<Frame>) -> Result<(), rav1e::EncoderStatus>;
+/*
+  fn receive_packet(&mut self) -> Packet;
+  fn container_sequence_header(&self) -> Vec<u8>;
+
+  fn twopass_bytes_needed(&self) -> usize;
+  fn twopass_in(&mut self, buf: &[u8]) -> Result<usize, EncoderStatus>;
+  fn twopass_out(&self) -> Vec<u8>;*/
+}
+
+impl EncoderContext for rav1e::Context<u16> {
+  fn new_frame(&self) -> Frame {
+    rav1e::Context::<u16>::new_frame(self).into()
+  }
+  fn send_frame(&mut self, frame: Option<Frame>) -> Result<(), rav1e::EncoderStatus> {
+    if let Some(frame) = frame {
+      if let Frame::U16(f) = frame {
+        self.send_frame(f)
+      } else {
+        Err(rav1e::EncoderStatus::Failure)
+      }
+    } else {
+      self.send_frame(None)
+    }
+  }
+}
+
+impl EncoderContext for rav1e::Context<u8> {
+  fn new_frame(&self) -> Frame {
+    rav1e::Context::<u8>::new_frame(self).into()
+  }
+  fn send_frame(&mut self, frame: Option<Frame>) -> Result<(), rav1e::EncoderStatus> {
+    if let Some(frame) = frame {
+      if let Frame::U8(f) = frame {
+        self.send_frame(f)
+      } else {
+        Err(rav1e::EncoderStatus::Failure)
```

Comment:
Won't this end up (once the other funcs are done) basically be two whole separate copied implementations?

That seems not ideal compared to having some sort of macro'd `match` inside each function, and having only *one* implementation...

Reply:
Ideally we can go with the same pattern I applied to `rav1e_frame_fill_plane()`, I wanted to try the dyn Trait way as well and see what works better all in all.

``` rust
match (ctx, frame) {
    (Context::U8(c), Frame::U8(f)) => foo_internal(...),
    (Context::U16(c), Frame::U16(f)) => foo_internal(...),
    _ => SomeError
}
```

Probably is compact enough

Reply:
My preference is the latter, yeah - much more compact and only in one location.